src/foolscap/slicers/root.py

```diff
@@ -54,22 +68,56 @@ def slicerForObject(self, obj):
             s = tokens.ISlicer(copier)
             return s
 
-        slicerFactory = self.slicerTable.get(type(obj))
+        slicerFactory = None
+        
+        # PY3KPORT
+        if six.PY2:
+            slicerFactory = self.slicerTable.get(type(obj))
+        elif six.PY3:
+            if type(obj) not in self.slicerTable:
+                # Determine if this is an instance
+                # Unfortunately this is True for file objects as well
+                # so we need another check as file objects have a different slicer
+                if hasattr(obj, '__class__') and isinstance(obj.__class__, type) and (not hasattr(obj, 'fileno')):
+                    # This is a class instance
+                    slicerFactory = self.slicerTable.get(InstanceType)
+
+            # Also for class methods i.e if obj == C.f where C is a
+            # class and f is its method, in Py3 the type evaluates
+            # to 'function' not 'method' whereas in Py2 it returns
+            # MethodSlicer, so we need to take care of that too.
+            elif type(obj) == types.FunctionType:
+                # Check the __qualname__ of the function - if it has
+                # a . then it means it belongs to a class so return
+                # MethodSlicer instead!
+                if hasattr(obj, '__qualname__') and '.' in obj.__qualname__:
+                    log.msg("{} is a method, not a function".format(obj))
+                    slicerFactory = self.slicerTable.get(types.MethodType)
+                else:
+                    slicerFactory = self.slicerTable.get(type(obj))
+                    
+            else:
+                slicerFactory = self.slicerTable.get(type(obj))
+            
```

Comment:
I'm not sure I understand what's going on here enough to be convinced we need completely different py2 vs py3 code? (Also this whole added block should go in its own function or method and have unit-tests).